pkg/chains/chain_filters.go

```diff
@@ -0,0 +1,60 @@
+package chains
+
+// ChainFilter is a function that filters chains based on some criteria
+type ChainFilter func(c Chain) bool
+
+// FilterExternalChains filters chains that are external
+func FilterExternalChains(c Chain) bool {
+	return c.IsExternal
+}
+
+// FilterGatewayObserver filters chains that are gateway observers
+func FilterGatewayObserver(c Chain) bool {
+	return c.CctxGateway == CCTXGateway_observers
+}
+
+// FilterConsensusEthereum filters chains that have the ethereum consensus
+func FilterConsensusEthereum(c Chain) bool {
+	return c.Consensus == Consensus_ethereum
+}
+
+// FilterConsensusBitcoin filters chains that have the bitcoin consensus
+func FilterConsensusBitcoin(c Chain) bool { return c.Consensus == Consensus_bitcoin }
+
+// FilterConsensusSolana filters chains that have the solana consensus
+func FilterConsensusSolana(c Chain) bool { return c.Consensus == Consensus_solana_consensus }
+
+// FilterChains applies a list of filters to a list of chains
+func FilterChains(chainList []Chain, filters ...ChainFilter) []Chain {
+	// Apply each filter to the list of supported chains
+	for _, filter := range filters {
```

Comment:
If we have 5 chains and 3 filters that match ALL chains, doesn't this mean that we'll have 15 chains with duplicates?

Reply:
What do you think about this?
```go
func FilterChains(items []Chain, filters ...ChainFilter) []Chain {
	// noop
	if len(filters) == 0 {
		return items
	}

	var result []Chain
	for _, item := range items {
	TOP:
		for _, filter := range filters {
			if !filter(item) {
				continue TOP
			}
		}

		result = append(result, item)
	}

	return result
}
```

Reply:
What do you mean ?
`chainList = filteredChains` we are not appending to the returned list . Does this clarify your question

Reply:
added a test case with duplicate filters

Reply:
Oh, okay, my bad. Still initial implementation would iterate though chains multiple times and do up to `len(chains) * len(filters)` allocations in the worst case. In proposed implementation it would do only up to `len(chains)`

Reply:
The length of chains is fixed , I wouldn't worry too much about the number of iterations and prioritize readabilty .
IMO using labels makes it less readable